File: python/gitProcess.py
import subprocess


class GitProcess:
    def __init__(self, repo) -> None:
        self.repo = repo
        self.pipe = subprocess.PIPE

    def Execute(self, args) -> str:
        process = subprocess.Popen(args, stdout=self.pipe, stderr=self.pipe)
        stdoutput, stderror = process.communicate()
        return "{}{}".format(stdoutput, stderror)

    def GetRemoteBranches(self) -> list:
        output = self.Execute(["git", "-C", self.repo, "branch", "-r"])
        branches = list(map(str.strip, output.__str__().replace("b'", "").lstrip().split("\\n")[:-1]))
        for branch in branches:
            branchName = branch.replace("origin/", "")
            output = self.Execute(
                ["git", "-C", self.repo, "branch", "--track", branchName, branch]
            )
        return branches

    def Pull(self, all: bool = True) -> str:
        return self.Execute(["git", "-C", self.repo, "pull", "--all" if all else ""])

    def Fetch(self, all: bool = True) -> str:
        return self.Execute(["git", "-C", self.repo, "fetch", "--all" if all else ""])

    def GetLogs(self, since, until, author = "") -> str:
        arguments = ["git", "-C", self.repo, "--no-pager", "log", "--since=\"{} 00:00:00\"".format(since)]
        if until:
            arguments.append("--until=\"{} 23:59:59\"".format(until))

        if author:
            arguments.append("--author={}".format(author))

        arguments.append('--numstat')
        arguments.append('--all')
        arguments.append('--source')
        output = self.Execute(arguments)
        return output

File: src/lib/gitProcess.ts
import { execSync } from 'child_process';
import * as vscode from 'vscode';

export class GitProcess {
    private repo: string;

    constructor(repo: string) {
        this.repo = repo;
    }

    private execute(args: string[]): string {
        const command = args.join(' ');

        try {
            const output = execSync(command, { encoding: 'utf-8' });
            return output;
        } catch (error) {
            vscode.debug.activeDebugConsole.appendLine(`Failed to execute command: ${command}`);
        }

        return '';
    }

    public getRemoteBranches(): string[] {
        const output = this.execute(['git', '-C', this.repo, 'branch', '-r']);
        const branches = output.split('\n').map(branch => branch.trim()).filter(branch => branch !== '');
        for (const branch of branches) {
            const branchName = branch.replace('origin/', '');
            this.execute(['git', '-C', this.repo, 'branch', '--track', branchName, branch]);
        }
        return branches;
    }

    public pull(all: boolean = true): string {
        return this.execute(['git', '-C', this.repo, 'pull', all ? '--all' : '']);
    }

    public fetch(all: boolean = true): string {
        return this.execute(['git', '-C', this.repo, 'fetch', all ? '--all' : '']);
    }

    public getLogs(since: string, until: string, author: string = ''): string {
        const args = ['git', '-C', this.repo, '--no-pager', 'log', `--since="${since} 00:00:00"`];
        if (until) {
            args.push(`--until="${until} 23:59:59"`);
        }
        if (author) {
            args.push(`--author=${author}`);
        }
        args.push('--numstat');
        args.push('--all');
        args.push('--source');
        const output = this.execute(args);
        return output;
    }
}

